### PR TITLE
hw1-cleancode

### PR DESCRIPTION
**Pull request by StasNikonov**
Add ESLint, Prettier, and GitHub Action for linting

### CODE REVIEW

**Comment by VfourTwenty** [comment]
### 1) The app doesn't throw the error if the port which it's supposed to use is busy. ###
   in index.js
 replaced this:
   app.listen(PORT, () => {
   console.log(`🚀 Сервер запущено на порті ${PORT}`);
   });

 with this:
app.listen(PORT, () => {
   console.log(`🚀 Сервер запущено на порті ${PORT}`);
   }).on('error', (err) => {
   console.error('❌ Failed to start server:', err.message);
   process.exit(1);
   });

this way an error message was shown when I had another program running on port 3000.


### 2) Request optimization: ###
Mailing service makes a call to the WeatherApi per each email being sent. 
This poses an issue when there are multiple cities (let's say 10000) subscribed for the same
frequency of updates in the same city. The actual forecast can be fetched once per city with 
1 api call and cashed in the db (create a new table for that) so that the data is looked up 
from cache and not refetched many times.
   
from sendWeatherEmailToSubscribers function:
   **for (let subscriber of subscribers) {**
   const { email, city, token } = subscriber;

   try {
   **const weather = await getWeather(city);**


### 3) Separating environments ###
 There are three different environments defined in config.js but they all read the same env variables.
 The same db is used for dev and testing, while its possible to use separate ones for clarity and simplicity. The test themselves check all the edge cases and are well handled.


### 4) Code style and readability ### 
Good idea with using Prettier. Code is really easy to read.

### 5) Modularity ###
I really like how logic, email templates, utilities etc are structured out in separate folders. And splitting routes and controllers I believe is the right approach, something I didn't consider.


#### Suggestions: ####
1) Optimize weather fetching logic
2) Add a script to seed the db
3) Add shortcuts for actions like migrating in package.json scripts

Good job overall!

**Review by darkravchuk — COMMENTED** [review]
I really like how you've separated the validation logic and email templates — this structure makes the code more organized, readable, and easier to maintain.
I recommend adding a .dockerignore file to optimize the Docker image build process.

File: src/api/controllers/subscriptionController.js [hunk 1]
```diff
@@ -1,19 +1,25 @@
...
 const subscribe = async (req, res) => {
   const { email, city, frequency } = req.body;
 
-  const validation = await validateSubscriptionInput({ email, city, frequency });
+  const validation = await validateSubscriptionInput({
+    email,
+    city,
+    frequency,
+  });
   if (!validation.valid) {
```
Comment (vasyl-khrapko-skelar) [inline src/api/controllers/subscriptionController.js:23]: Хорошою практикою є вставляти пустий рядок перед логічними блоками (if, for і т.д.), що підвищує читабельність коду

File: src/api/middlewares/cityValidator.js [hunk 2]
```diff
@@ -1,17 +1,18 @@
 const axios = require('axios');
```
Comment (vasyl-khrapko-skelar) [inline src/api/middlewares/cityValidator.js:1]: У файлах з директорії middlewares ти зібрав логіку валідації, але ці функції не є middleware, доцільніше було б їх винести в функції utils (це можна зробити в наступних дз)

File: src/utils/forecastTemplates.js [hunk 3]
```diff
@@ -1,4 +1,10 @@
-const forecastEmail = ({ city, temp_c, humidity, condition, unsubscribeLink }) => `
+const forecastEmail = ({
```
Comment (vasyl-khrapko-skelar) [inline src/utils/forecastTemplates.js:1]: За Clean Code краще використовувати action verbs в назві функцій (в цьому випадку, наприклад, buildForecastEmail)

```suggestion
const buildForecastEmail = ({
```

File: src/db/config/config.js [hunk 4]
```diff
@@ -24,5 +24,5 @@ module.exports = {
     host: process.env.DB_HOST,
     port: process.env.DB_PORT,
     dialect: 'postgres',
-  }
+  },
```
Comment (vasyl-khrapko-skelar) [inline src/db/config/config.js:27]: розкажеш чому обрав саме такий підхід в роботі з конфігом?
Загалом тут дублюються змінні конфіги, які будуть менеджитись на рівні різних enviroments (різних .env файлів), тому можна було зробити через один об'єкт для отримання змінних, але різні env конфіги)

**Review by vasyl-khrapko-skelar — APPROVED** [review]
Хороша робота

Залишив мінорні коменти по чистоті коду, також очікую, що в дз#3 з SOLID/GRASP буде змінена структура компонентів проєкту, є питання до розбиття на слої та відповідальності класів, звернені до 3rd party (API погоди) напряму через axios, а не через клієнт
